bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java

```diff
@@ -0,0 +1,207 @@
+            final String clientPrincipalName = clientKerberosName.toString();
+            LOG.debug("Using JAAS/SASL/GSSAPI auth to connect to server Principal {}", serverPrincipal);
+            try {
+                saslClient = Subject.doAs(clientSubject, new PrivilegedExceptionAction<SaslClient>() {
+                    @Override
+                    public SaslClient run() throws SaslException {
+                        String[] mechs = {"GSSAPI"};
+                        return Sasl.createSaslClient(mechs, clientPrincipalName, serviceName, serviceHostname, null,
+                            new ClientCallbackHandler(null));
+                    }
+                });
+            } catch (PrivilegedActionException err) {
+                LOG.error("GSSAPI client error", err.getCause());
+                throw new SaslException("error while booting GSSAPI client", err.getCause());
+            }
+        }
+        if (saslClient == null) {
+            throw new SaslException("Cannot create JVM SASL Client");
+        }
+
+    }
+
+    public byte[] evaluateChallenge(final byte[] saslToken) throws SaslException {
+        if (saslToken == null) {
+            throw new SaslException("saslToken is null");
+        }
+        if (clientSubject != null) {
+            try {
+                final byte[] retval
+                    = Subject.doAs(clientSubject, new PrivilegedExceptionAction<byte[]>() {
+                        @Override
+                        public byte[] run() throws SaslException {
+                            return saslClient.evaluateChallenge(saslToken);
+                        }
+                    });
+                return retval;
+            } catch (PrivilegedActionException e) {
+                LOG.error("SASL error", e.getCause());
+                throw new SaslException("SASL/JAAS error", e.getCause());
+            }
+        } else {
+            return saslClient.evaluateChallenge(saslToken);
+        }
+    }
+
+    private Subject loginClient(ClientConfiguration clientConfiguration) throws SaslException, LoginException {
+        boolean systemRole = ClientConfiguration.CLIENT_ROLE_SYSTEM.equals(clientConfiguration.getClientRole());
+        String configurationEntry = systemRole
+            ? clientConfiguration.getString(SaslConstants.JAAS_AUDITOR_SECTION_NAME, SaslConstants.JAAS_DEFAULT_AUDITOR_SECTION_NAME)
+            : clientConfiguration.getString(SaslConstants.JAAS_CLIENT_SECTION_NAME, SaslConstants.JAAS_DEFAULT_CLIENT_SECTION_NAME);
+        AppConfigurationEntry[] entries = Configuration.getConfiguration()
+            .getAppConfigurationEntry(configurationEntry);
+        if (entries == null) {
+            LOG.info("No JAAS Configuration found with section BookKeeper");
+            return null;
+        }
+        try {
+            LoginContext loginContext = new LoginContext(configurationEntry, new ClientCallbackHandler(null));
```

Comment:
I have checked my code.
My implementation creates a new LoginContext (in SASLBookieAuthProvider/SaslServerState) for every incoming connection so there is no need to perform a TGT renewal, in fact I'm not running in such a problem in "staging".
I see it can be a performance/resource problem, in by staging env I have a limited number of clients (10 to 100) and I have never got into problems.

The same happens at the client-side.

@revans2 
do you think we must cache the TGT as ZooKeeper ?